File: LC/[0126] Word Ladder II.cpp
class Solution {
public:
    unordered_set<string>wordSet;
    unordered_map<string,vector<string>>next;
    unordered_map<string,unordered_set<string>>prev;
    vector<vector<string>>results;
    string beginWord;

    vector<vector<string>> findLadders(string beginWord, string endWord, vector<string>& wordList)
    {
        wordList.push_back(beginWord);

        //create word_set
        for (auto str:wordList)
            wordSet.insert(str);
        this->beginWord = beginWord;

        //edge case: if end word not in word_list: just return empty list
        if (wordSet.find(endWord)==wordSet.end()) return {};

        //construct connected word graph
        for (int i=0; i<wordList.size(); i++)
        {
            string str = wordList[i];
            for (int k=0; k<str.size(); k++)
            {
                for (char ch='a'; ch<='z'; ch++)
                {
                    string temp = str;
                    temp[k] = ch;
                    if (temp==str) continue;
                    if (wordSet.find(temp)!=wordSet.end())
                        next[str].push_back(temp);
                }
            }
        }

        unordered_set<string>visited;
        queue<string>q;
        int flag = 0;

        //BFS to find the valid path
        q.push(beginWord);
        visited.insert(beginWord);

        while (!q.empty())
        {
            int len = q.size();
            unordered_set<string>newVisited;

            for (int k=0; k<len; k++)//for each level
            {
                string str = q.front();
                q.pop();

                for (auto nextStr: next[str])//add all children for next level
                {
                    if (visited.find(nextStr)!=visited.end()) //if already visited
                        continue;
                    prev[nextStr].insert(str);//update ancestor mapping table
                    newVisited.insert(nextStr);

                    if (nextStr==endWord) flag = 1;
                }
            }

            for (auto x: newVisited)
            {
                visited.insert(x);
                q.push(x);
            }

            if (flag==1) break;
        }

        if (flag==0) return {};

        DFS(endWord,{endWord});

        return results;
    }

    void DFS(string word, vector<string> path)
    {
        if (word==beginWord)
        {
            reverse(path.begin(),path.end());
            results.push_back(path);
            return;
        }

        for (auto preStr:prev[word])
        {
            vector<string>newPath=path;
            newPath.push_back(preStr);
            DFS(preStr, newPath);
        }
        return;
    }
};

File: LC/[0126] Word Ladder II.py
from collections import defaultdict, deque
class Solution:
    def findLadders(self,
                     beginWord: str,
                     endWord: str,
                     wordList: List[str]) -> int:

        # build the graph
        g=defaultdict(set)
        word_set=set([beginWord]+wordList)
        if endWord not in word_set: return []
        my_ancestors=defaultdict(set)
        #print("word_set :", word_set)
        for w in [beginWord]+wordList:
            #print(" adding w: ", w)
            wl=list(w)
            for ii in range(len(wl)):
                for jj in range(26):
                    cc=wl[ii]
                    nc=chr(ord('a')+jj)
                    if nc==cc: continue
                    wl[ii]=nc
                    nw="".join(wl)
                    wl[ii]=cc
                    if nw in word_set:
                        g[w].add(nw)
        #print(g)
        # BFS to count the levels reaching the ending word
        dq=deque([beginWord])
        visited=set()
        lvl=0
        is_found=False
        while (sz:=len(dq)) > 0:
            lvl+=1
            for _ in range(sz):
                cw=dq.popleft()

                if cw==endWord:
                    is_found=True
                    break

                if cw in visited: continue
                visited.add(cw)
                for child in g[cw]:
                    my_ancestors[child].add(cw)
                    dq.append(child)

            if is_found: break

        if not is_found: lvl=0
        #print(my_ancestors)
        paths=[]
        path=[]
        visited=set()
        def dfs(curr, path):
            if curr==beginWord:
                if (len(path)+1) == lvl:
                    paths.append([curr]+path)
                return

            if len(path)>= lvl: return

            for parent in my_ancestors[curr]:
                if parent in visited: continue
                visited.add(parent)

                dfs(parent, [curr]+path)
                visited.remove(parent)

        #print("lvl: ", lvl)
        dfs(endWord, path)

        return paths
